### PR TITLE
luminous: librbd: image create request should validate data pool for self-managed snapshot support

### PR DESCRIPTION
**Pull request by smithfarm**
http://tracker.ceph.com/issues/24946

### CODE REVIEW

**Comment by smithfarm** [comment]
jenkins test docs

**Comment by smithfarm** [comment]
```
/home/jenkins-build/build/workspace/ceph-pull-requests/src/librbd/image/CreateRequest.cc: In member function ‘void librbd::image::CreateRequest<ImageCtxT>::validate_data_pool()’:
/home/jenkins-build/build/workspace/ceph-pull-requests/src/librbd/image/CreateRequest.cc:245:19: error: ‘m_io_ctx’ was not declared in this scope
   m_data_io_ctx = m_io_ctx;
                   ^
src/librbd/CMakeFiles/rbd_internal.dir/build.make:734: recipe for target 'src/librbd/CMakeFiles/rbd_internal.dir/image/CreateRequest.cc.o' failed
make[3]: *** [src/librbd/CMakeFiles/rbd_internal.dir/image/CreateRequest.cc.o] Error 1
```

**Comment by smithfarm** [comment]
```
/home/jenkins-build/build/workspace/ceph-pull-requests/src/librbd/image/CreateRequest.cc: In member function ‘void librbd::image::CreateRequest<ImageCtxT>::validate_data_pool()’:
/home/jenkins-build/build/workspace/ceph-pull-requests/src/librbd/image/CreateRequest.cc:255:41: error: ‘class librados::IoCtx’ has no member named ‘get_namespace’
     m_data_io_ctx.set_namespace(m_ioctx.get_namespace());
                                         ^
src/librbd/CMakeFiles/rbd_internal.dir/build.make:734: recipe for target 'src/librbd/CMakeFiles/rbd_internal.dir/image/CreateRequest.cc.o' failed
make[3]: *** [src/librbd/CMakeFiles/rbd_internal.dir/image/CreateRequest.cc.o] Error 1
```

**Comment by smithfarm** [comment]
The problem here is that, in master, `m_io_ctx` is declared in `src/librbd/image/CreateRequest.h` like this:

```
IoCtx m_io_ctx;
```

while in luminous the `m_ioctx` declaration looks like this:

```
IoCtx &m_ioctx;
```

**Comment by smithfarm** [comment]
So, my guess is the commit in this PR needs to be cherry-picked after 9c59e688eac

**Comment by smithfarm** [comment]
jenkins test docs

**Comment by smithfarm** [comment]
@trociny @dillaman I got it to compile, but it blows up in the unit tests. Any suggestions welcome, including closing the PR :-)

File: src/librbd/image/CreateRequest.cc [hunk 1]
```diff
@@ -277,27 +263,29 @@ void CreateRequest<I>::validate_pool() {
...
+void CreateRequest<I>::handle_validate_data_pool(int r) {
   ldout(m_cct, 20) << "r=" << r << dendl;
 
-  if (r == 0) {
-    validate_overwrite();
+  bufferlist bl;
+  bl.append("overwrite validated");
+
+  if (r >= 0 && m_outbl.contents_equal(bl)) {
+    add_image_to_directory();
```
Comment (trociny) [inline src/librbd/image/CreateRequest.cc]: @smithfarm Note, in luminus CreateRequest state machine is a little different than in the master (see the diagram in CreateRequest.h). In luminous the next state after VALIDATE_DATA_POOL is CREATE_ID_OBJECT, so you need to call `create_id_object()` here instead.

**Comment by smithfarm** [comment]
Thanks, @trociny, it passes make check now. Ready for review :-)

**Comment by yuriw** [comment]
[wip-yuri-testing-2018-10-03-2001-luminous](https://trello.com/c/WouIZtJT/433-wip-yuri-testing-2018-10-03-2001-luminous)

**Pull request merged**